### DISCUSSION TITLE
Transient prompt in Xonsh shell

### DISCUSSION THREAD

**Discussion by ahillio** [xonsh/xonsh#4702] [General]
I'm so excited about the possibility of switching to xonsh! I installed it, testdriving it, enabled vi-mode... and now... I'm facing that this "transient prompt" feature is really a necessity for me...

Transient prompt allows you to have a highly detailed prompt without all the details cluttering up the command history in your terminal. You can see it here https://raw.githubusercontent.com/romkatv/powerlevel10k-media/master/transient-prompt.gif -- notice how clean and peaceful the terminal looks without `~/repo master +1` and so forth prefacing every command?

Can xonsh do this?

**Comment by dyuri** [comment]
Interesting concept, I like it!
I don't think that it can be achieved with the current codebase and I'm not a `prompt_toolkit` expert, but I'd definitely use it if it would be available.

**Reply by jnoortheen** [reply]
well we can change the session.prompt attribute . I have used it for the AsyncPrompt. We can do the same here I guess

**Reply by ahillio** [reply]
I don't know what that means exactly but it sounds hopeful...
Does this mean it's currently possible?
Or does this mean it will have to be added as a new feature?

**Reply by jnoortheen** [reply]
something like this 
<img width="712" alt="image" src="https://user-images.githubusercontent.com/6702219/157584732-24056a3d-0e1c-48e9-ade4-ebbbb57336ba.png">

off the bat, it doesn't update multi-line prompt, can't update the prompt after entering commands. It needs to be added to upstream ptk project

**Comment by macintacos** [comment]
I know that this thread is a bit old, but I'm moving my configuration over from `fish` and I missed my transient prompt. So I came up with the following solution that seems to work quite well for my purposes:

```python
# EXPERIMENTAL transient prompt. Ported intent from fish: user/prompt.fish
# (fish_prompt --final-rendering + fish_right_prompt timestamp).
# Collapses the submitted prompt to "$ command" with a right-aligned timestamp.
# Primary mechanism: return the condensed prompt while prompt_toolkit is in its
# final post-accept render (app.is_done). Layered on top of 50-prompt.xsh.
# Disable by moving this file out of rc.d/.
import datetime
from prompt_toolkit.application.current import get_app_or_none

# REQUIRED: without this, xonsh computes the prompt once (static) and never
# re-evaluates $PROMPT on the final post-accept render, so the is_done branch
# below can never fire. With it on, `message` is a callable ptk re-evaluates
# each render (fields are still cached per-prompt, so this isn't per-keystroke
# git work). See xonsh/shells/ptk_shell/__init__.py: UPDATE_PROMPT_ON_KEYPRESS.
$UPDATE_PROMPT_ON_KEYPRESS = True

_FULL_PROMPT = $PROMPT
_FULL_RIGHT = $RIGHT_PROMPT if "RIGHT_PROMPT" in __xonsh__.env else ""

def _dynamic_prompt():
    app = get_app_or_none()
    if app is not None and app.is_done:
        return "$ "
    return _FULL_PROMPT
$PROMPT = _dynamic_prompt

def _dynamic_right_prompt():
    app = get_app_or_none()
    if app is not None and app.is_done:
        return datetime.datetime.now().strftime("[%b %d %H:%M:%S]")
    return _FULL_RIGHT
$RIGHT_PROMPT = _dynamic_right_prompt
```

This is what it looks like on my machine:

<img width="2098" height="390" alt="CleanShot 2026-06-05 at 21 15 45@2x" src="https://github.com/user-attachments/assets/3be9356b-77a3-4519-8b0c-94420e9d4266" />

Adjust to your liking. Last thing to call out - [`$UPDATE_PROMPT_ON_KEYPRESS`](https://xon.sh/envvars.html#UPDATE_PROMPT_ON_KEYPRESS) says that there's a performance penalty. I can't really tell what that performance penalty is, but YMMV.

**Comment by bestlem** [comment]
An alternative is to put the details of the prompt at the bottom of the terminal.
See  $BOTTOM_TOOLBAR in general.

xontrib_starship can be set to use starship as the bottom toolbar

**Comment by uninstall-your-browser** [comment]
As an alternative to the above, it's simple enough to update the prompt after entering commands (i.e. right before they are actually ran), also without `$UPDATE_PROMPT_ON_KEYPRESS`
```python
# ~/.xonshrc
from datetime import datetime

from prompt_toolkit import PromptSession, Application
from xonsh.built_ins import XSH
from xonsh.shells.ptk_shell import PromptToolkitShell
from xonsh.events import events

$MINIMAL_PROMPT = "{BOLD_BLUE}{prompt_end}{RESET} "
$MINIMAL_RPROMPT = lambda: datetime.now().strftime("[%b %d %H:%M:%S]")

@events.on_ptk_create
def register(prompter: PromptSession, **kwargs):
    def minimise_prompt_after_command_run(app: Application):
        if app.is_done:
            shell: PromptToolkitShell = XSH.shell.shell
            prompter.message = shell._get_prompt_tokens("MINIMAL_PROMPT", "minimal_prompt")
            prompter.rprompt = shell._get_prompt_tokens("MINIMAL_RPROMPT", "minimal_rprompt", default=[])

    prompter.app.before_render.add_handler(minimise_prompt_after_command_run)
```
In fact, you can change the prompt any time you want, as long as you call `app.renderer.erase()` after - this is done by the function that fires `before_render` for us here.

Example:
<img width="1124" height="91" alt="image" src="https://github.com/user-attachments/assets/8cf28942-d1cb-43a1-b7f9-9f97ffa01a7b" />
